Source/WebCore/style/ElementRuleCollector.cpp

```diff
@@ -559,8 +559,25 @@ void ElementRuleCollector::collectMatchingRulesForList(const RuleSet::RuleDataVe
 
         auto addRuleIfMatches = [&] (const Element* scopingRoot = nullptr) {
             unsigned specificity;
-            if (ruleMatches(ruleData, specificity, matchRequest.styleScopeOrdinal, scopingRoot))
-                addMatchedRule(ruleData, specificity, matchRequest);
+            if (ruleMatches(ruleData, specificity, matchRequest.styleScopeOrdinal, scopingRoot)) {
+                auto computeDistance = [&] {
+                    constexpr auto maximumDistance = std::numeric_limits<unsigned>::max();
+                    if (!scopingRoot)
+                        return maximumDistance;
+
+                    unsigned distance = 0;
+                    auto current = &element();
+                    while (current) {
+                        if (current == scopingRoot)
+                            return distance;
+                        current = current->parentElement();
+                        ++distance;
+                    }
+                    ASSERT_NOT_REACHED(); // The scoping root must exist at this point.
+                    return distance;
+                };
```

Comment:
Why don't we compute this distance above at the same time we loop to find the scoping roots?

This should be factored into a function. Let's not to put special case things like this into `collectMatchingRulesForList` directly.